File: pocs/code-review-demo/agents/perf_agent.py
from beeai_framework.agents.experimental import RequirementAgent
from beeai_framework.memory import UnconstrainedMemory
from beeai_framework.tools.think import ThinkTool
from core.backend import get_llm


def create_perf_agent() -> RequirementAgent:
    return RequirementAgent(
        llm=get_llm(),
        memory=UnconstrainedMemory(),
        tools=[ThinkTool()],
        role="performance_reviewer",
        instructions=(
            "You are a performance engineer reviewing code efficiency. "
            "Focus on: algorithmic complexity, nested loops, N+1 queries, and unnecessary allocations. "
            "Cross-reference any findings against prior notes in memory. "
            "Report each issue with a severity: info, warning, or critical."
        ),
    )

File: pocs/code-review-demo-ts/src/agents/perfAgent.ts
import { RequirementAgent } from "beeai-framework/agents/requirement/agent";
import { ConditionalRequirement } from "beeai-framework/agents/requirement/requirements/conditional";
import { UnconstrainedMemory } from "beeai-framework/memory/unconstrainedMemory";
import { ThinkTool } from "beeai-framework/tools/think";
import { getLLM } from "../core/backend.js";

export async function createPerfAgent() {
  return new RequirementAgent({
    llm: await getLLM(),
    memory: new UnconstrainedMemory(),
    tools: [new ThinkTool()],
    instructions:
      "You are a performance engineer. Review the submitted code for algorithmic complexity issues, " +
      "inefficient loops, unnecessary re-computation, and N+1 query patterns. " +
      "Return a structured list of findings with severity: CRITICAL, WARNING, or INFO.",
    requirements: [
      new ConditionalRequirement(ThinkTool, { forceAtStep: 1, maxInvocations: 3 }),
    ],
  });
}
